math/src/msm/mod.rs

```diff
@@ -0,0 +1,174 @@
+use crate::cyclic_group::IsCyclicBilinearGroup;
+use crate::field::element::FieldElement;
+use crate::field::traits::IsLinearField;
+
+pub mod naive;
+pub mod pippenger;
+
+pub trait MSM<F, G>
+where
+    G: IsCyclicBilinearGroup,
+    F: IsLinearField,
+{
+    fn msm(&self, ks: &[FieldElement<F>], ps: &[G]) -> G;
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::elliptic_curve::curves::test_curve::ORDER_P;
+    use crate::elliptic_curve::element::EllipticCurveElement;
+    use crate::elliptic_curve::traits::{HasDistortionMap, HasEllipticCurveOperations};
+    use crate::field::element::FieldElement;
+    use crate::field::fields::u64_prime_field::U64PrimeField;
+    use crate::field::quadratic_extension::{HasQuadraticNonResidue, QuadraticExtensionField};
+
+    use super::pippenger;
+    use super::*;
+    use test_case::test_case;
+
+    #[derive(Debug, Clone)]
+    pub struct QuadraticNonResidue;
+    impl HasQuadraticNonResidue<U64PrimeField<ORDER_P>> for QuadraticNonResidue {
+        fn residue() -> FieldElement<U64PrimeField<ORDER_P>> {
+            -FieldElement::one()
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct CurrentCurve;
+    impl HasEllipticCurveOperations for CurrentCurve {
+        type BaseField = QuadraticExtensionField<U64PrimeField<ORDER_P>, QuadraticNonResidue>;
+
+        fn a() -> FieldElement<Self::BaseField> {
+            FieldElement::from(1)
+        }
+
+        fn b() -> FieldElement<Self::BaseField> {
+            FieldElement::from(0)
+        }
+
+        fn generator_affine_x() -> FieldElement<Self::BaseField> {
+            FieldElement::from(35)
+        }
+
+        fn generator_affine_y() -> FieldElement<Self::BaseField> {
+            FieldElement::from(31)
+        }
+
+        fn embedding_degree() -> u32 {
+            2
+        }
+
+        fn order_r() -> u64 {
+            5
+        }
+
+        fn order_p() -> u64 {
+            59
+        }
+    }
+
+    impl HasDistortionMap for CurrentCurve {
+        fn distorsion_map(
+            p: &[FieldElement<Self::BaseField>; 3],
+        ) -> [FieldElement<Self::BaseField>; 3] {
+            let (x, y, z) = (&p[0], &p[1], &p[2]);
+            let t = FieldElement::new([FieldElement::zero(), FieldElement::one()]);
+            [-x, y * t, z.clone()]
+        }
+    }
```

Comment:
I've seen a lot of this code be moved somewhere in some draft PR (which sounds good); for now leaving it where it was.